File: server/src/datasources/job.js
const { RESTDataSource } = require('apollo-datasource-rest');
const parseStringPromise = require('xml2js').parseStringPromise;

class JobAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = 'http://stackoverflow.com/jobs/';//?q=as&l=london&u=Km&d=20';
  }

  async getJobs() {
    const response = await this.get('feed');
    const converted = await parseStringPromise(response, { explicitArray: false, ignoreAttrs: true});
    const items = converted.rss.channel.item;
    console.log('Total Jobs:', converted.rss.channel['os:totalResults'])
    return Array.isArray(items)
      ? items.map(job => this.jobReducer(job))
      : [];
  }

  jobReducer(job) {
    return {
      id: job.guid || 0,
      cursor: `${new Date(job.pubDate).getTime()}`,
      title: job.title,
      description: job.description,
      publishDate: new Date(job.pubDate),
      category: Array.isArray(job.category) ? job.category : [job.category],
      location: job.location,
      link: job.link,
      company: job['a10:author']['a10:name'],
    };
  }
}

module.exports = JobAPI;

File: client/lib/job.dart
import 'package:intl/intl.dart';

class Job {
  String id;
  String title;
  String description;
  String company;
  String link;
  DateTime publishDate;
  String location;
  List<String> category;

  Job({
    this.id,
    this.title,
    this.description,
    this.company,
    this.link,
    this.publishDate,
    this.location,
    this.category,
  });

  factory Job.fromMap(Map<String, dynamic> map) {
    return Job(
      id: map['id'] as String,
      title: map['title'] as String,
      description: map['description'] as String,
      company: map['company'] as String,
      link: map['link'] as String,
      publishDate:
          DateTime.fromMillisecondsSinceEpoch(map['publishDate'] as int),
      location: map['location'] as String,
      category: map['category'].cast<String>(),
    );
  }
}
